smarts/core/scenario.py

```diff
@@ -115,15 +115,13 @@ class TraverseGoal(Goal):
     the vehicle must be going the correct direction in its lane
     just prior to doing so."""
 
-    def __init__(self, road_network: SumoRoadNetwork):
-        super().__init__()
-        self._road_network = road_network
+    road_network: SumoRoadNetwork
```

Comment:
While I am slightly confused about having an initialiser in a frozen dataclass I do not think we should be exposing the `road_network` through this object.

Reply:
I agree, what would be a better way? Keeping the initializer with the frozen attribute throws an error. Removing the frozen attribute breaks Python 3.8 since it inherits from a frozen dataclass. We could remove frozen on both this class and the `Goal` class, but that would break immutability and hashing (we would have to add `unsafe_hash=True`).

Since we just use the `road_network` for `drove_off_map()` perhaps we could just pass a closure that wraps that function to avoid passing the whole network?

Reply:
Yes, I think the problem here I am seeing is that we are not using a `dataclass` not as just data. The `road_network` is not serialisable and I would be somewhat wary of providing any reference value to a `dataclass`. The solution I can think of is forcing `is_reached()` to also take the `road_network` as input but I have some fear that will require the `road_network` in places it should not show up. Perhaps that fear is unfounded though. 

Otherwise we could consider completely stripping the `dataclass` decorator. I think a large part of the purpose of the `core.scenario` versions of the goals were to operate as objects rather than just data or else it is not worth much to spend the effort converting from `sstudio` to `core.scenario` scenarios.